cajita/src/Cajita_HypreSemiStructuredSolver.hpp

```diff
@@ -0,0 +1,1565 @@
+    void setMaxIter( const int max_iter ) { this->setMaxIterImpl( max_iter ); }
+
+    //! Set the output level.
+    void setPrintLevel( const int print_level )
+    {
+        this->setPrintLevelImpl( print_level );
+    }
+
+    //! Set a preconditioner.
+    void
+    setPreconditioner( const std::shared_ptr<HypreSemiStructuredSolver<
+                           Scalar, EntityType, MemorySpace>>& preconditioner )
+    {
+        // This function is only valid for non-preconditioners.
+        if ( _is_preconditioner )
+            throw std::logic_error(
+                "Cannot call setPreconditioner() on a preconditioner" );
+
+        // Only a preconditioner can be used as a preconditioner.
+        if ( !preconditioner->isPreconditioner() )
+            throw std::logic_error( "Not a preconditioner" );
+
+        _preconditioner = preconditioner;
+        this->setPreconditionerImpl( *_preconditioner );
+    }
+
+    //! Setup the problem.
+    void setup()
+    {
+        // This function is only valid for non-preconditioners.
+        if ( _is_preconditioner )
+            throw std::logic_error( "Cannot call setup() on preconditioners" );
+
+        auto error = HYPRE_SStructMatrixAssemble( _A );
+        checkHypreError( error );
+
+        this->setupImpl( _A, _b, _x );
+    }
+
+    /*!
+      \brief Solve the problem Ax = b for x.
+      \param b The forcing term.
+      \param x The solution.
+    */
+    template <class Array_t>
+    void solve( const Array_t& b, Array_t& x, int n_vars = 3 )
+    {
+        Kokkos::Profiling::pushRegion( "Cajita::HypreSemiStructuredSolver::solve" );
+
+        static_assert( is_array<Array_t>::value, "Must use an array" );
+        static_assert(
+            std::is_same<typename Array_t::entity_type, entity_type>::value,
+            "Array entity type mush match solver entity type" );
+        static_assert(
+            std::is_same<typename Array_t::memory_space, MemorySpace>::value,
+            "Array device type and solver device type are different." );
+
+        static_assert(
+            std::is_same<typename Array_t::value_type, value_type>::value,
+            "Array value type and solver value type are different." );
+
+        // This function is only valid for non-preconditioners.
+        if ( _is_preconditioner )
+            throw std::logic_error( "Cannot call solve() on preconditioners" );
+
+        // Spatial dimension.
+        const std::size_t num_space_dim = Array_t::num_space_dim;
+
+        int part = 0;
+
+        // Initialize the RHS.
+        auto error = HYPRE_SStructVectorInitialize( _b );
+        checkHypreError( error );
+
+        // Copy the RHS into HYPRE. The HYPRE layout is fixed as layout-right.
+        auto owned_space = b.layout()->indexSpace( Own(), Local() );
+        std::array<long, num_space_dim + 1> reorder_size;
+        for ( std::size_t d = 0; d < num_space_dim; ++d )
+        {
+            reorder_size[d] = owned_space.extent( d );
```

Comment:
We'll need to get the `min()` and `max()` and use both to construct these index spaces - using just extent will likely look the same as the previous result

Reply:
The index space setup has been altered, but it and the entire handling of the views for setting the HYPRE vectors (LHS and RHS) within `solve` need to be looked over. The current format works, but there is likely a way to optimize it